### PR TITLE
Overview documentation overhaul

### PR DESCRIPTION
**Pull request by marcgarreau**
### What was wrong?
The existing [Overview](https://web3py.readthedocs.io/en/stable/overview.html) documentation provides an introduction to Providers, then dumps out the base class API, e.g., utility methods for type conversions, address helpers, etc.

### How was it fixed?
I thought the Overview page would be put to better use (i.e., answer more typical support questions) if it instead gave a tour of the features in the lib. I believe this will be useful for orienting newcomers, but I also built the reference page I wished existed; I included much of the API, broken into usage category. 

There's a lot of content and its a big departure from the current state of things. Feedback appreciated!

Generated docs preview: https://web3py--1681.org.readthedocs.build/en/1681/

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

### CODE REVIEW

File: docs/overview.rst [hunk 1]
```diff
@@ -1,380 +1,306 @@
...
 
-    Takes a variety of inputs and returns its string equivalent.
-    Text gets decoded as UTF-8.
 
+Your Keys
+~~~~~~~~~
 
-    .. code-block:: python
+Private keys are required to approve any transaction made on your behalf. The type of
+provider you use will determine how you interact with those keys and will affect the
```
Comment (pipermerriam) [inline docs/overview.rst:71]: The provider type doesn't actually dictate this.  Maybe a more accurate statement would be:

> The manner in which your key is secured will determine how you create and send transactions in Web3.py.

File: docs/overview.rst [hunk 2]
```diff
@@ -1,380 +1,306 @@
...
+- :meth:`Web3.keccak() <web3.Web3.keccak>`
+- :meth:`Web3.solidityKeccak() <web3.Web3.solidityKeccak>`
 
-    Returns ``True`` if the value is a valid `EIP55`_ checksummed address
 
+web3.eth API
+~~~~~~~~~~~~
 
-    .. code-block:: python
+Most of this library's functionality can be found under the ``web3.eth`` namespace.
```
Comment (pipermerriam) [inline docs/overview.rst:127]: Maybe reword as:

> The most commonly used APIs for interacting with Ethereum can be found under ...

File: docs/overview.rst [hunk 3]
```diff
@@ -1,380 +1,306 @@
...
-        '0xd3CdA913deB6f67967B99D67aCDFa1712C293601'
 
-.. _EIP55: https://github.com/ethereum/EIPs/issues/55
+Making Transactions
+-------------------
 
+The most common use cases will be satisfied with
+:meth:`sendTransaction <web3.eth.Eth.sendTransaction>` or the combination of
+:meth:`signTransaction <web3.eth.Eth.signTransaction>` and
+:meth:`sendRawTransaction <web3.eth.Eth.sendRawTransaction>`. If interacting
```
Comment (pipermerriam) [inline docs/overview.rst:163]: The *"If interacting with smart contracts..."* part seems like it belongs on it's own line (maybe as a `.. note::`)

File: docs/overview.rst [hunk 4]
```diff
@@ -1,380 +1,306 @@
...
 
-    Returns the Keccak-256 as it would be computed by the solidity ``keccak``
-    function on the provided ``value`` and ``abi_types``.  The ``abi_types``
-    value should be a list of solidity type strings which correspond to each
-    of the provided values.
+The two most common use cases involving smart contracts are deploying and executing
+functions on a deployed contract.
 
+Deployment requires that the contract already be compiled, with its bytecode and ABI
+available. This compilation step is typically done within
```
Comment (pipermerriam) [inline docs/overview.rst:193]: Maybe take out the `typically` and reword with language like *"can be done with ..."*

File: docs/overview.rst [hunk 5]
```diff
@@ -1,380 +1,306 @@
...
+If you want to read data from a contract (or see the result of transaction locally,
+without executing it on the network), you can use the :meth:`ContractFunction.call`
+method:
 
-.. py:classmethod:: Web3.sha3(primitive=None, hexstr=None, text=None)
+.. code-block:: python
 
-    .. WARNING::
-      This method has been deprecated for :meth:`~Web3.keccak`
+   >>> deployed_contract.functions.getMyValue().call()
```
Comment (pipermerriam) [inline docs/overview.rst:222]: This might be a good place to use the `Contract.caller` API which allows for slighly more concise syntax.

File: docs/overview.rst [hunk 6]
```diff
@@ -1,380 +1,306 @@
...
 
-Check Encodability
-~~~~~~~~~~~~~~~~~~~~
+- :meth:`web3.eth.filter() <web3.eth.Eth.filter>`
+- :meth:`web3.eth.getFilterChanges() <web3.eth.Eth.getFilterChanges>`
+- :meth:`web3.eth.getFilterLogs() <web3.eth.Eth.getFilterLogs>`
+- :meth:`web3.eth.uninstallFilter() <web3.eth.Eth.uninstallFilter>`
+- :meth:`web3.eth.getLogs() <web3.eth.Eth.getLogs>`
+- :meth:`Contract.events.your_event_name.createFilter() <web3.contract.Contract.events.your_event_name.createFilter>`
+- :meth:`Contract.events.your_event_name.build_filter() <web3.contract.Contract.events.your_event_name.build_filter>`
```
Comment (pipermerriam) [inline docs/overview.rst:276]: Maybe also add `your_event.processReceipt`, `your_event.processLog`, `your_event.getLogs`

File: docs/overview.rst [hunk 7]
```diff
@@ -1,380 +1,306 @@
...
-Modules
--------
+ENS
+~~~
 
-The JSON-RPC functionality is split across multiple modules which *loosely*
-correspond to the namespaces of the underlying JSON-RPC methods.
+`Ethereum Name Service (ENS) <https://ens.domains/>`_ provides the infrastructure
+for human-readable addresses. As an example, instead of
+``0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045``, you can send funds to ``vitalik.eth``.
```
Comment (carver) [inline docs/overview.rst:305]: *nit* Super minor: name-checking Vitalik very subtly feeds into a (false) narrative that eth devs worship him, or that he wields some kind of dictatorial power. I would use `ethereumfoundation.eth` or something like it instead.

File: docs/overview.rst (hunk 5 above)
Comment (marcgarreau) [reply docs/overview.rst:222]: I wasn't even familiar with this API 😅. Will include both routes.

File: docs/overview.rst (hunk 6 above)
Comment (marcgarreau) [reply docs/overview.rst:286]: Looks like a small ticket worth of work to add a `getLogs` method to the events documentation, plus sort out links for `processLog` and `processReceipt`. The latter two get lumped under the same top-level link without links of their own ([ContractEvents.myEvent](https://web3py.readthedocs.io/en/stable/contracts.html#web3.contract.ContractEvents.myEvent)). Will look to sort that out in a subsequent PR, so I can start this beta release.
